segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorageManager.java

```diff
@@ -0,0 +1,817 @@
+                                                                   return zkClient.delete().forPath(getZkPath(streamSegmentName) + "/" + child).toCompletableFuture()
+                                                                                  .thenApply(v -> null);
+                                                               }
+                                                               ledger.addToList(offset, ledgerData);
+                                                               CompletableFuture<LedgerData> retval = new CompletableFuture<LedgerData>();
+                                                               retval.complete(ledgerData);
+                                                               return retval;
+                                                           });
+                                               })
+                                               .toArray(CompletableFuture[]::new);
+                                       return CompletableFuture.allOf(futures).thenApply(v -> {
+                                           return ledger;
+                                       });
+                                   }
+                               });
+    }
+
+    private void tryDeleteLedger(long ledgerId) {
+        bookkeeper.asyncDeleteLedger(ledgerId, (rc, ctx) -> {
+            if (rc != 0) {
+                log.warn("Deletion of ledger {} failed with errorCode {}", ledgerId, rc);
+            }
+        }, null);
+    }
+
+    private CompletableFuture<LedgerData> fenceLedgerAt(String streamSegmentName, int firstOffset) {
+        Stat stat = new Stat();
+        return zkClient.getData()
+                       .storingStatIn(stat)
+                       .forPath(getZkPath(streamSegmentName) + "/" + firstOffset)
+                       .toCompletableFuture()
+                       .thenApply(data -> {
+                           LedgerData ledgerData = deserializeLedgerData(firstOffset, data, false, stat);
+                           ledgerData.setLength((int) ledgerData.getLedgerHandle().getLength());
+                           ledgerData.setLastAddConfirmed(ledgerData.getLedgerHandle().getLastAddConfirmed());
+                           try {
+                               ledgerData.getLedgerHandle().close();
+                           } catch (Exception e) {
+                               throw new CompletionException(e);
+                           }
+                           return ledgerData;
+                       });
+    }
+
+
+    private CompletableFuture<Void> writeDataAt(LedgerHandle lh, long offset, InputStream data, int length, String segmentName) {
+        log.info("Writing {} at offset {} for ledger {}", length, offset, lh.getId());
+        byte[] bytes = new byte[length];
+        CompletableFuture<Void> retVal = new CompletableFuture<>();
+
+        try {
+            StreamHelpers.readAll(data, bytes, 0, length);
```

Comment:
this method returns a number which tells you how many bytes were read. You should check that to make sure you read as much as you think you did.